src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs

```diff
@@ -1413,7 +1416,112 @@ private IBlockOperation CreateBoundBlockOperation(BoundBlock boundBlock)
+                    unaryOperatorMethod: null,
+                    _semanticModel,
+                    syntax,
+                    compilation.GetSpecialType(SpecialType.System_Boolean),
+                    constantValue,
+                    isImplicit: true);
+            }
+
+            var argumentNullExceptionMethodSymbol = (IMethodSymbol)compilation.GetWellKnownTypeMember(WellKnownMember.System_ArgumentNullException__ctorString);
+            var argumentNullExceptionType = compilation.GetWellKnownType(WellKnownType.System_ArgumentNullException);
+            var argumentNullExceptionObject = new ObjectCreationOperation(
+                        constructor: argumentNullExceptionMethodSymbol,
+                        initializer: new ObjectOrCollectionInitializerOperation(
+                                initializers: ImmutableArray<IOperation>.Empty,
+                                _semanticModel,
+                                syntax,
+                                compilation.GetWellKnownType(WellKnownType.System_ArgumentNullException),
+                                constantValue,
+                                isImplicit: true),
+                        ImmutableArray.Create<IArgumentOperation>(
+                            new ArgumentOperation(
+                                new LiteralOperation(_semanticModel, syntax, compilation.GetSpecialType(SpecialType.System_String), parameter.Name, isImplicit: true),
+                                ArgumentKind.Explicit,
+                                parameter: argumentNullExceptionMethodSymbol.Parameters[0],
+                                inConversionOpt: null,
+                                outConversionOpt: null,
+                                _semanticModel,
+                                syntax,
+                                isImplicit: true)),
+                        _semanticModel,
+                        syntax,
+                        type: argumentNullExceptionType,
+                        constantValue,
+                        isImplicit: true);
+
+            IOperation whenTrue = new ExpressionStatementOperation(
+                new ThrowOperation(
+                    argumentNullExceptionObject,
+                    _semanticModel,
+                    syntax,
+                    type: argumentNullExceptionType,
+                    constantValue,
+                    isImplicit: true),
+                _semanticModel,
+                syntax,
+                compilation.GetSpecialType(SpecialType.System_Void),
+                constantValue,
+                isImplicit: true);
+            return new ConditionalOperation(conditionOp, whenTrue, null, isRef: false, _semanticModel, syntax, compilation.GetSpecialType(SpecialType.System_Boolean), constantValue, isImplicit: true);
+
+        }
+
+        private ImmutableArray<ConditionalOperation> GenerateNullChecksForParameters(ImmutableArray<ParameterSymbol> parameters, SyntaxNode syntax)
```

Comment:
>ImmutableArray<ConditionalOperation> [](start = 16, length = 36)

Should these `IOperation` nodes be generated lazily? (I don't know what the guidance is for when nodes are generated lazily.)

Reply:
I don't think it's worth generating these lazily. We could, certainly, but it would require adding quite a bit of machinery as there's no corresponding bound node for the if checks. These nodes don't have unbounded recursion, and while functions with large parameter lists will generate a big list of `if`s, the cost seems pretty small to me.

---
In reply to: [305538085](https://github.com/dotnet/roslyn/pull/37381#discussion_r305538085) [](ancestors = 305538085)

Reply:
Less machinery sounds good to me.

---
In reply to: [305548063](https://github.com/dotnet/roslyn/pull/37381#discussion_r305548063) [](ancestors = 305548063,305538085)